### ISSUE TITLE
False-positive Cannot call method on

### ISSUE THREAD

**Issue by fluffycondor** [phpstan/phpstan-doctrine#152]
# Bug report

PHPStan 0.12.50

`Cannot call method update() on App\Entity\MyEntity|false` when it can't be false.

### Code snippet that reproduces the problem

```php
/** @var \Doctrine\Common\Collections\Collection<MyEntity> $collection */
if ($collection->isEmpty()) {
    $collection->add($new);
} else {
    $collection->first()->update($new); // Cannot call method update() on App\Entity\MyEntity|false
}
```
### Expected output

There should not be any error, but Cannot call method update() on App\Entity\MyEntity|false raised

**Comment by ondrejmirtes** [comment]
This is not a bug but a well-known limitation described in the documentation: https://phpstan.org/writing-php-code/narrowing-types

This can be solved with a simple TypeSpecifyingExtension (https://phpstan.org/developing-extensions/type-specifying-extensions). When `isEmpty()` is truthy, `first()` and other methods should return `null`. When it's falsey, `null` should be removed from `first()` and other methods typehints.

**Comment by fluffycondor** [comment]
@ondrejmirtes I've tried to write a TypeSpecifyingExtension for the case based on some examples I've found in phpstan/phpstan-symfony, but it doesn't work as I expected. Everything I got - PHPStan gives me a new error
`Call to method Doctrine\Common\Collections\Collection<int,App\Entity\MyEntity>::isEmpty() will always evaluate to false.`, and `Cannot call method on` errors are still there.

Could you please explain what am I doing wrong?

This is my code:
```php
<?php

declare(strict_types=1);

namespace App\PHPStan;

use PhpParser\Node\Expr\MethodCall;
use PHPStan\Analyser\Scope;
use PHPStan\Analyser\SpecifiedTypes;
use PHPStan\Analyser\TypeSpecifier;
use PHPStan\Analyser\TypeSpecifierAwareExtension;
use PHPStan\Analyser\TypeSpecifierContext;
use PHPStan\Broker\Broker;
use PHPStan\Reflection\MethodReflection;
use PHPStan\Reflection\ParametersAcceptorSelector;
use PHPStan\Type\BooleanType;
use PHPStan\Type\MethodTypeSpecifyingExtension;
use PHPStan\Type\TypeCombinator;

final class DoctrineCollectionFirstTypeSpecifyingExtension implements MethodTypeSpecifyingExtension, TypeSpecifierAwareExtension
{
    private const COLLECTION_CLASS = 'Doctrine\Common\Collections\Collection';
    private const IS_EMPTY_METHOD_NAME = 'isEmpty';
    private const FIRST_METHOD_NAME = 'first';

    private Broker $broker;
    private TypeSpecifier $typeSpecifier;

    public function __construct(Broker $broker)
    {
        $this->broker = $broker;
    }

    public function getClass(): string
    {
        return self::COLLECTION_CLASS;
    }

    public function isMethodSupported(
        MethodReflection $methodReflection,
        MethodCall $node,
        TypeSpecifierContext $context
    ): bool {
        return $methodReflection->getName() === self::IS_EMPTY_METHOD_NAME && $context->false();
    }

    public function specifyTypes(
        MethodReflection $methodReflection,
        MethodCall $node,
        Scope $scope,
        TypeSpecifierContext $context
    ): SpecifiedTypes {
        $classReflection = $this->broker->getClass(self::COLLECTION_CLASS);
        $methodVariants = $classReflection->getNativeMethod(self::FIRST_METHOD_NAME)->getVariants();

        return $this->typeSpecifier->create(
            new MethodCall($node->var, self::FIRST_METHOD_NAME),
            TypeCombinator::remove(new BooleanType(), ParametersAcceptorSelector::selectSingle($methodVariants)->getReturnType()),
            $context
        );
    }

    public function setTypeSpecifier(TypeSpecifier $typeSpecifier): void
    {
        $this->typeSpecifier = $typeSpecifier;
    }
}
```

**Comment by ondrejmirtes** [comment]
Please send that as PR so I can do a proper review.

**Comment by ruudk** [comment]
Should be fixed by https://github.com/phpstan/phpstan-doctrine/pull/172

**Comment by ondrejmirtes** [comment]
Thank you :)
